### PR TITLE
Initial OARO implementation

### PR DESCRIPTION
**Pull request by adam-a-a**
## Summary/Motivation:
Thanks goes out to @luohezhiming for pair programming sessions.

This PR implements a bilateral counter-current RO module for osmotically assisted RO. 0D implementation is in place and is working for calculated CP effects and mass transfer coefficient. 

Subsequent PR(s) for additional refinement will be put up, but wanted to get this initial model up first. TODO notes have been put in place to serve as a reminder for refinement to be made in subsequent PR(s).

## Changes proposed in this PR:
- add osmotically_assisted_reverse_osmosis base and 0D files
- refine membrane channel as needed

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/core/membrane_channel0d.py [hunk 1]
```diff
@@ -114,21 +116,29 @@ def add_state_blocks(self, has_phase_equilibrium=None):
                     },
                 },
             )
-        else:
+        elif self._flow_direction == FlowDirection.backward:
+            # TODO: added this more explicit conditional as a sanity check,
+            # but it seems the intended referencing below does not work, i.e.,
+            # properties_in[0] == properties[0, 0] (what's actually happening)
+            # instead of properties_in[0] == properties[0, 1] (intended but not happening)
```
Comment (adam-a-a) [inline watertap/core/membrane_channel0d.py]: @bknueven Unsure why this didn't work for the permeate channel which is setup to have backward flow direction. However, taking this code outside and putting directly in oaro_0D after `add_state_blocks` ending up creating the intended object references. For now, I have the intended functionality working, but would like you to take a look.

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_0D.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                width_var=self.width,
+                flow_direction=flow_direction,
+            )
+            if not hasattr(self, "eq_area"):
+                add_eq_area = True
+            else:
+                add_eq_area = False
+            self._add_area(include_constraint=add_eq_area)
+        else:
+            mem_side.add_geometry(length_var=None, width_var=None)
```
Comment (bknueven) [inline watertap/unit_models/osmotically_assisted_reverse_osmosis_0D.py]: @adam-a-a is the issue with `MembraneChannel0DBlock` that the `flow_direction` is not passed in here?

File: watertap/core/membrane_channel_base.py [hunk 3]
```diff
@@ -429,7 +429,7 @@ def eq_cp_modulus(b, t, x, j):
                 self.length_domain,
                 solute_set,
                 initialize=5e-5,
-                bounds=(1e-6, 1e-3),
+                bounds=(1e-10, 1e-3),
```
Comment (adam-a-a) [inline watertap/core/membrane_channel_base.py]:
```suggestion
                bounds=(1e-6, 1e-3),
```

File: watertap/core/membrane_channel_base.py [hunk 4]
```diff
@@ -398,7 +398,7 @@ def eq_cp_modulus(b, t, x, j):
             self.length_domain,
             solute_set,
             initialize=1.1,
-            bounds=(0.1, 3),
+            bounds=(0, 3),
```
Comment (adam-a-a) [inline watertap/core/membrane_channel_base.py]:
```suggestion
            bounds=(0.1, 3),
```

File: watertap/core/membrane_channel_base.py [hunk 5]
```diff
@@ -568,7 +568,7 @@ def _add_calculated_pressure_change_mass_transfer_components(self):
 
         self.spacer_porosity = Var(
             initialize=0.95,
-            bounds=(0.1, 0.99),
+            # bounds=(0.1, 0.99),
```
Comment (adam-a-a) [inline watertap/core/membrane_channel_base.py]:
```suggestion
            bounds=(0.1, 0.99),
```

**Review by adam-a-a — COMMENTED** [review]
Committing suggestions to revert changes in bounds

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_0D.py (hunk 2 above)
Comment (adam-a-a) [reply watertap/unit_models/osmotically_assisted_reverse_osmosis_0D.py]: Seems like a good thing that we should add although not sure whether it'll serve as the remedy.

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_0D.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            self.config.property_package.component_list,
+            doc="Mass transfer from feed to permeate",
+        )
+        def eq_connect_mass_transfer(b, t, p, j):
+            return (
+                b.permeate_side.mass_transfer_term[t, p, j]
+                == -b.feed_side.mass_transfer_term[t, p, j]
+            )
+
+        # ------------------------------------------------------------------------------------------------
```
Comment (tarnold17) [inline watertap/unit_models/osmotically_assisted_reverse_osmosis_0D.py]: Is there any reason this comment block can't just be removed?

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py [hunk 7]
```diff
@@ -0,0 +1,818 @@
...
+# Technology Laboratory (subject to receipt of any required approvals from
+# the U.S. Dept. of Energy). All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. These files are also available online at the URL
+# "https://github.com/watertap-org/watertap/"
+#
+###############################################################################
+
+from copy import deepcopy
```
Comment (tarnold17) [inline watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py]: `deepcopy`, `Enum`, and `auto` appear to be unused - can these import statements be removed?

File: watertap/core/membrane_channel0d.py (hunk 1 above)
Comment (luohezhiming) [inline watertap/core/membrane_channel0d.py]: I want to confirm if this problem has been resolved by your modification? I am also confused why this not happen.

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py [hunk 8]
```diff
@@ -0,0 +1,818 @@
...
+        #                     blk.conc_mass_phase_comp["Liq", j]
+        #                 )
+        #             if blk.is_property_constructed("mole_frac_phase_comp"):
+        #                 self._rescale_permeate_variable(blk.mole_frac_phase_comp[j])
+        #             if blk.is_property_constructed("molality_phase_comp"):
+        #                 self._rescale_permeate_variable(
+        #                     blk.molality_phase_comp["Liq", j]
+        #                 )
+        #         if blk.is_property_constructed("pressure_osm_phase"):
+        #             self._rescale_permeate_variable(blk.pressure_osm_phase["Liq"])
```
Comment (luohezhiming) [inline watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py]: Probably you can delete these comments if they are redundant.

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py [hunk 9]
```diff
@@ -0,0 +1,818 @@
...
+
+        for v in self.rejection_phase_comp.values():
+            if iscale.get_scaling_factor(v) is None:
+                iscale.set_scaling_factor(v, 1)
+
+        if hasattr(self, "structural_parameter"):
+            if iscale.get_scaling_factor(self.structural_parameter) is None:
+                # Structural parameter expected to be ~ 1200 microns
+                iscale.set_scaling_factor(self.structural_parameter, 1e3)
+        # if not hasattr(self, "_permeate_scaled_properties"):
```
Comment (tarnold17) [inline watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py]: Is there any reason this comment block can't be removed?

**Review by tarnold17 — APPROVED** [review]
It looks okay to me, I left a few minor comments. One other question I have - I see that tests have only been added for `osmotically_assisted_reverse_osmosis_0D.py`. Do tests also need to be added for `osmotically_assisted_reverse_osmosis_1D.py` and `osmotically_assisted_reverse_osmosis_base.py`?

**Review by MarcusHolly — APPROVED** [review]
I'm not going to say I perfectly grasp everything that has been done here, but for the most part, everything seems logical and consistent to me. To piggyback off Travis's comment - are the missing test files something that Chenyu, Travis, and I will work on in the subsequent PR(s)?

**Review by luohezhiming — COMMENTED** [review]
Looks great! I am glad that OARO0D model works when all configuration set `calculated` and the 1D model is also added. The tests for OARO1D cases might be added later!

**Comment by adam-a-a** [comment]
@MarcusHolly @tarnold17 @luohezhiming Thanks for your reviews! You pointed out something important that completely slipped my mind. I only created the 1d file to get started with a template but never went back to fix it up. So, we'll want to get the 1D version in too. We'll also want to get a flowsheet going as well. 

As for this PR, the only thing I can solidly stand by is the implementation of the 0D unit with tests. Perhaps I'll delete the 1d file now so that coverage check passes, we'll merge this, then huddle together for next steps.

File: watertap/core/membrane_channel0d.py (hunk 1 above)
Comment (adam-a-a) [reply watertap/core/membrane_channel0d.py]: I think it was resolved by passing the flow_direction into the appropriate method (add_geometry I think it was).

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py (hunk 9 above)
Comment (adam-a-a) [reply watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py]: I cleaned this up.

File: watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py (hunk 7 above)
Comment (adam-a-a) [reply watertap/unit_models/osmotically_assisted_reverse_osmosis_base.py]: Removing those unused imports

File: watertap/core/membrane_channel0d.py (hunk 1 above)
Comment (bknueven) [reply watertap/core/membrane_channel0d.py]: Yep, that was the issue.

**Pull request merged**